keps/sig-auth/2579-pod-isolation-policy/README.md

```diff
@@ -0,0 +1,951 @@
+
+### Non-Goals
+
+1. Provide a configurable policy mechanism that meets the needs of all use-cases
+2. Limit privilege escalation and other attacks beyond the pod to host boundary
+   (e.g. policies on services, secrets, etc.)
+
+
+## Proposal
+
+The three profile levels (privileged, baseline, restricted) of the [Pod Security Standards] will
+be hardcoded into the new admission plugin. Changes to the standards will be tied to the Kubernetes
+version that they were introduced in.
+
+### API
+
+<<[UNRESOLVED]>>
+
+The annotation prefix `podisolationpolicy.kubernetes.io` is open to discussion. The proposed
+prefix structure is `<featurename>.kubernetes.io`, and depends on the final name we settle on.
+
+<<[/UNRESOLVED]>>
+
+Policy application is controlled based on labels on the namespace. The following labels are supported:
+```
+podisolationpolicy.kubernetes.io/allow: <policy level>
+podisolationpolicy.kubernetes.io/allow-version: <policy version>
+podisolationpolicy.kubernetes.io/audit: <policy level>
+podisolationpolicy.kubernetes.io/audit-version: <policy version>
+podisolationpolicy.kubernetes.io/warn: <policy level>
+podisolationpolicy.kubernetes.io/warn-version: <policy version>
+```
+
+**Allow:** Pods meeting the requirements of the allowed level are allowed. Violations are rejected
+in admission.
+
+**Audit:** Pods and [templated pods] meeting the requirements of the audit
+policy level are ignored. Violations are recorded in a
+`podisolationpolicy.kubernetes.io/audit: <violation>` annotation on the audit event for the request.
+
+**Warn:** Pods and [templated pods] meeting the requirements of the warn
+level are ignored. Violations are returned in a user-facing warning message.
+
+[templated pods]: #podtemplate-resources
+
+<<[UNRESOLVED]>>
+
+The mode key names are open to discussion. In particular, I'm not satisfied that `allow` and
+`audit/warn` carry slightly different meanings:
+
+- `allow` means "allow pods meeting this profile level & below" (i.e. deny above)
+- `audit` and `warn` mean "audit/warn on pods exceeding this profile level" (i.e. audit/warn above)
+
+Using `deny` for the enforcing level would be more consistent with audit/warn (deny above), but is a
+counter-intuitive user experience. One would expect `deny = privileged` to deny privileged pods, but
+it really means allow privileged pods (hence using `allow` for the key).
+
+Another option is to change the `audit` and `warn` keys to something meaning "don't audit/warn at this
+level and bellow", but we haven't thought of a good name. The best alternative we've come up with is
+`audit-above` and `warn-above`, but that is cumbersome.
+
+A final option is to change `allow` to `enforce`, which is more ambiguous. The ambiguity makes the
+inconsistency with audit & warn less obvious, but doesn't help the user experience.
+
+<<[/UNRESOLVED]>>
+
+### Validation
+
+The following restrictions are placed (by the admission plugin) on the policy namespace labels:
```

Comment:
why enforced via admission as opposed to enforcing in the kube validation code itself since we have direct access to do so?  If we don't enforce all the time, we could have bad values leak in over time, right?

Reply:
keeping the validation code for the mechanism in a single place is easier to understand, allows it to be completely disabled along with the admission plugin if someone doesn't want any part of this mechanism active in their cluster, and would allow it to be run as part of an admission webhook on an older cluster